client-core/src/wallet/syncer.rs

```diff
@@ -547,9 +547,31 @@ impl<
     }
 }
 
+/// testnet v0.5
+const GENESIS_APP_HASH: &str = "F62DDB49D7EB8ED0883C735A0FB7DE7F2A3FA322FCD2AA832F452A62B38607D5";
+
+fn check_genesis_hash(genesis_hash: &[u8]) -> Result<()> {
+    let hash_setted = std::env::var("GENESIS_APP_HASH").unwrap_or_else(|_| GENESIS_APP_HASH.into());
+    let genesis_hash_str = String::from_utf8(genesis_hash.to_vec()).chain(|| {
+        (
+            ErrorKind::VerifyError,
+            "genesis app hash is invalid from tendermint",
+        )
+    })?;
+    if hash_setted == genesis_hash_str {
+        Ok(())
+    } else {
+        let msg = format!(
+            "genesis app hash from tendermint {} is not match app hash {}",
+            genesis_hash_str, hash_setted
+        );
+        Err(Error::new(ErrorKind::VerifyError, msg))
+    }
+}
+
 pub fn get_genesis_sync_state<C: Client>(client: &C) -> Result<SyncState> {
-    // FIXME verify against trusted genesis hash
     let genesis = client.genesis()?;
+    check_genesis_hash(&genesis.app_hash)?;
```

Comment:
It's not enough to check app hash of genesis, but the hash of whole genesis.

Reply:
tendermint gives the genesis, if the app_hash is wrong in genesis, tendermint can not start up.

Reply:
> tendermint gives the genesis, if the app_hash is wrong in genesis, tendermint can not start up.

We can't trust the whole genesis response from the node, and app hash only covers the app state, don't cover other important informations, like genesis_time, consensus_params, validators.

Reply:
I think `app_hash` does cover genesis time and consensus parameters. I may be wrong.

Reply:
> I think `app_hash` does cover genesis time and consensus parameters. I may be wrong.

No, it only covers `ChainState` plus valid tx ids in the last block.

Reply:
yes, I'll the whole genesis app hash instead of app hash

Reply:
https://github.com/crypto-com/chain/blob/ae5271ca1ebc9bf8fd24bd340cf828a9a2e76ed1/chain-abci/src/app/app_init.rs#L224